### PR TITLE
HCL Expression AST

### PR DESCRIPTION
**Pull request by lkishalmi**
This one adds the AST for the HCL Expression language. That would be required for a more in-depth semantic analyzer, formatter, hints, whatever.

### CODE REVIEW

File: ide/languages.hcl/nbproject/project.properties [hunk 1]
```diff
@@ -14,5 +14,6 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
-javac.source=1.8
+javac.source=11
+javac.target=11
```
[the remark this answers is not in the document]
Comment (lkishalmi) [reply ide/languages.hcl/nbproject/project.properties:18]: That solved the failing test issues!

[the remark this answers is not in the document]
Comment (mbien) [reply ide/languages.hcl/nbproject/project.properties:18]: cool. going to add the all-tests label to exclude surprises in the graal/lsp tests. We can remove it again if it was green once.

File: ide/languages.hcl/manifest.mf [hunk 2]
```diff
@@ -3,4 +3,5 @@ OpenIDE-Module: org.netbeans.modules.languages.hcl
 OpenIDE-Module-Layer: org/netbeans/modules/languages/hcl/layer.xml
 OpenIDE-Module-Localizing-Bundle: org/netbeans/modules/languages/hcl/Bundle.properties
 OpenIDE-Module-Specification-Version: 1.1
+OpenIDE-Module-Java-Dependencies: Java > 11
```
Comment (vieiro) [inline ide/languages.hcl/manifest.mf:6]: Is this `>11` correct? Should it be `Java >= 11` instead?

File: ide/languages.hcl/src/org/netbeans/modules/languages/hcl/ast/expression/HCLExpressionFactory.java [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+                HCLArithmeticOperation.Operator op = binOp(ctx.op.getType());
+                return new HCLArithmeticOperation.Binary(op, expr(ctx.left), expr(ctx.right));
+            }
+            if (ctx.right != null) {
+                switch (ctx.op.getType()) {
+                    case NOT:
+                        return new HCLArithmeticOperation.Unary(HCLArithmeticOperation.Operator.NOT, expr(ctx.right));
+                    case MINUS:
+                        return new HCLArithmeticOperation.Unary(HCLArithmeticOperation.Operator.MINUS, expr(ctx.right));
+                }
```
Comment (vieiro) [inline ide/languages.hcl/src/org/netbeans/modules/languages/hcl/ast/expression/HCLExpressionFactory.java:61]: You may want to add some defensive programming here with a `default: throw new UnsupportedOperationException` here?

**Review by vieiro — APPROVED** [review]
Wow, this is looking great. I've left some comments you may want to review. Good job!

File: ide/languages.hcl/manifest.mf (hunk 2 above)
Comment (mbien) [reply ide/languages.hcl/manifest.mf:6]: it looks weird, but is correct :)
https://github.com/search?q=repo%3Aapache%2Fnetbeans+OpenIDE-Module-Java-Dependencies%3A+Java+%3E+11&type=code

I believe there is no `>=`. `>` here basically means greater equals

File: ide/languages.hcl/src/org/netbeans/modules/languages/hcl/ast/expression/HCLExpressionFactory.java (hunk 3 above)
Comment (lkishalmi) [reply ide/languages.hcl/src/org/netbeans/modules/languages/hcl/ast/expression/HCLExpressionFactory.java:61]: Could be. That would catch a change in the parser which would not have reflected back here.

**Comment by lkishalmi** [comment]
I would call for more review and approval once this PR would be ready. I've raised this as a preview to get help from @mbien resolbing my Java 11 vs UnitTest issues.

I would have liked this included in NetBeans 19. Though it is not ready yet, there are lot of crufts to be removed/polished, also this alone does not add any functionality to the IDE at the moment, Though many advanced IDE things requires to be able to have the complete AST available..

**Comment by lkishalmi** [comment]
image

Having better AST, the Semantic analyzer recognize, functions, object keys as well.

The coloring are not final, though it's time to move on the indexing, code completion and schema validation....

**Comment by lkishalmi** [comment]
@mbien I can wait for that PR and rebase.

Maybe including HCL support should not be in the base IDE...

Coding all of these I'm eager to move to at least Java 17...

**Comment by lkishalmi** [comment]
@mbien any news on the Graal? Moss is growing fat on this PR.

**Comment by mbien** [comment]
@lkishalmi #6369 is not ready yet and I had no time to take another look after the initial attempt.

its another case where dependencies where not kept up to date for a while.

**Comment by lkishalmi** [comment]
Rebased on master, squashed.  I hope for a green one, then would merge.

**Comment by mbien** [comment]
the graal job might need some restarts if tests fail since it is missing the retry script. I am going to add it again in a followup PR.

**Pull request merged**